### PR TITLE
Feature/vector search

### PR DESCRIPTION
**Pull request by acruikshank-wm**
## Problem

This PR introduces semantic (neural, vector) search to Testbench via a VectorSearchChain. Vector search allows the chain to retrieve text results from a database by their similarity to a provided query.

This feature relies on an external vector database that stores embeddings (currently Pinecone is the only supported database). It also requires access to an embedding model (which is somewhat different from a standard LLM model). This feature currently only supports OpenAI models and Huggingface models through the sentence-transformers package.

## Solution

This was already a pretty big PR, which I made worse by performing a few refactors. I'm adding comments to individual files to explain why they were changed or added and what sort of review is required.

1. Introduce VectorSearchChain as a new LangChain component chain that computes the embedding of a query and finds results from a vector database based on similarity to that embedding.
2. Add dependencies for pinecone and sentence-transformers that permits this chain to work.
3. Introduce VectorSearchSpec and designer UI to allow users to specify a vector search.
4. Refactor designer components so each component is in its own file.
5. Change styles to use sass and use style nesting to ensure styles are scoped to components.
6. Misc changes to ease python debugging

## Screenshots

Screen Shot 2023-07-27 at 1 09 14 PM

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
+{
```
Comment (acruikshank-wm) [inline .vscode/launch.json:1]: Added to ease python debugging.

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (acruikshank-wm) [inline .vscode/settings.json:1]: Added to ease python debugging

File: app/src/components/Designer.tsx
Comment (acruikshank-wm) [inline app/src/components/Designer.tsx:1]: This file had too much logic. I broke it out into individual designer components in the designers directory.

File: app/src/components/TextModal.tsx [hunk 3]
```diff
@@ -23,7 +23,7 @@ const TextModal = ({enterValue, title, buttonText, placeholder, validateInput, m
   }, [modalOpen, activeModalId]);
 
   const handleEnter = useCallback(async () => {
-    if (validateInput(text)) {
+    if (await validateInput(text)) {
```
Comment (acruikshank-wm) [inline app/src/components/TextModal.tsx:26]: This was a bug. validateInput was acting asynchronously, but had the wrong type.

File: app/src/components/designers/APISpecDesigner.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
+import { useContext, useEffect, useCallback, useState } from "react";
```
Comment (acruikshank-wm) [inline app/src/components/designers/APISpecDesigner.tsx:1]: This and all the Designer components other than VectorSearch are just a refactor to put this code in individual files. They require light review at most.

File: app/src/components/designers/DeleteChainButton.tsx [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import QuickMenu from "../QuickMenu"
```
Comment (acruikshank-wm) [inline app/src/components/designers/DeleteChainButton.tsx:1]: This is also just extracting code from the Designer component file.

File: app/src/components/designers/VectorSearchSpecDesigner.tsx [hunk 6]
```diff
@@ -0,0 +1,109 @@
+import { useContext, useEffect, useMemo, useState } from "react";
```
Comment (acruikshank-wm) [inline app/src/components/designers/VectorSearchSpecDesigner.tsx:1]: This is a new component and requires a more thorough review.

File: cli.py [hunk 7]
```diff
@@ -224,6 +224,8 @@ def once(chain_name, input, record):
   input_json = buffer.decode('utf-8')
   input = json.loads(input_json)
 
+  chain_service.initialize_services()
```
Comment (acruikshank-wm) [inline cli.py:227]: ensure we're initializing api keys.

File: requirements.txt
Comment (acruikshank-wm) [inline requirements.txt:1]: Upgraded LangChain and adding Pinecone, SentenceTransformers and their dependencies.

File: scripts/generate_openai_embeddings.py
Comment (acruikshank-wm) [inline scripts/generate_openai_embeddings.py:1]: This script helps set up a vector database with OpenAI embeddings.

File: scripts/generate_sentencetransformers_embeddings.py
Comment (acruikshank-wm) [inline scripts/generate_sentencetransformers_embeddings.py:1]: This script helps set up a vector database with Huggingface embeddings.

File: scripts/run_vector_search_chain.py
Comment (acruikshank-wm) [inline scripts/run_vector_search_chain.py:1]: This is some handy example code to test the VectorSearchChain.

File: scripts/send_to_pinecone.py
Comment (acruikshank-wm) [inline scripts/send_to_pinecone.py:1]: This is a utility to help populate a Pinecone database.

File: app/src/components/designers/ParameterEditor.tsx [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+const ParameterEditor: React.FC<ParameterEditorProps> = ({ editorKey, parameters, onChange }) => {
+  const [entries, setEntries] = useState<[string, string][]>(Object.entries(parameters));
+
+  useEffect(() => {
+    onChange(Object.fromEntries(entries));
+  }, [entries]);
+
+  const updateEntry = (idx: number, key: string, value: string) => (
+    setEntries([...entries.slice(0,idx), [key, value], ...entries.slice(idx+1)])
+  )
```
Comment (pdavis68) [inline app/src/components/designers/ParameterEditor.tsx:19]: Theoretically React's batched updates could cause this to mess up as updates from both onChange events could be batched together, if my understanding is correct.

I think it's pretty unlikely, though. You'd have to put something in both fields pretty quickly,

**Review by pdavis68 — CHANGES_REQUESTED** [review]
Minor, minor issues only.

File: app/src/components/designers/ParameterEditor.tsx (hunk 8 above)
Comment (acruikshank-wm) [reply app/src/components/designers/ParameterEditor.tsx:23]: I see what you're saying. I split this into two functions so at least there won't be trouble between the closures and the event handlers. Beyond that, I don't think can come up with a better strategy that doesn't produce a render loop.

**Review by okay-son — APPROVED** [review]
I think this is a very cool feature! I think the PR can(?) be approved from my understanding and review.

**Review by pdavis68 — APPROVED** [review]
Looks good. There's still a lot of error checking that ought to happen as this gets built up. For example, the use of environment variables without checking to see if they exist or are valid.  There are tons of embedded strings that, if I were doing C#, I would make constants.

For example, look at all the hard coded strings here:

```
  def vector_store(self, query) -> VectorStore:    
    if self.database == 'pinecone':
      if 'index' not in self.database_params:
        raise ValueError('Missing index parameter for Pinecone database')

      # import just-in-time so auth will happen after env vars are loaded
      import pinecone
      from langchain.vectorstores.pinecone import Pinecone
      
      index = pinecone.Index(self.database_params['index'])
      return Pinecone(
        index,
        query,
        self.database_params.get('text_key') or 'text',
        namespace=self.database_params.get('namespace') or ''
      )
    else:
      raise ValueError(f'Unknown database: {self.database}')


  def _call(self, inputs: Dict[str, str]) -> Dict[str, str]:
    if self.embedding_engine == 'openai':
      self.embedding_params['openai_api_key'] = os.environ.get("OPENAI_API_KEY")
      embeddings = OpenAIEmbeddings(**self.embedding_params)
    elif self.embedding_engine == 'huggingface':
      model = self.embedding_params.get('model_name') or 'all-mpnet-base-v2'
      embeddings = HuggingFaceEmbeddings(**self.embedding_params, model_name=model)
    else:
      raise ValueError(f'Unknown embedding engine: {self.embedding_engine}')
```

But that stuff is more long-term changes. But thought I should mention it while it's on my mind.

**Pull request merged**